memory viz: Add colors for categories and a legend

zdevito opened the pull request:
Adds a category legend to memory trace plots that colors allocations by their role (activation, parameter, gradient, etc.) as captured by kineto.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95753
* #95541
* #95357
* #95356
* #91336
* __->__ #90587

Differential Revision: [D43757381](https://our.internmc.facebook.com/intern/diff/D43757381)

zdevito committed:
memory viz: Add colors for categories and a legend

[ghstack-poisoned]

zdevito commented:
![Screen Shot 2022-12-09 at 4 20 55 PM](https://user-images.githubusercontent.com/370202/206818070-18862260-0cfd-45a0-9873-68502088048c.png)

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

robieta commented on torch/cuda/_memory_viz.py:
```diff
@@ -430,11 +463,11 @@ def profile_plot(memory_profile, device=None):
-            category = 'unknown'
-        else:
-            category = category.name.lower()
+            category = Category.TEMPORARY
```
Why the change? Just because something is uncategorized doesn't mean it's `TEMPORARY`.

zdevito replied:
I didn't know what to categorize something without a category. Should we make an unknown or uncategorized category in the legend?

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"




[ghstack-poisoned]

aaronenyeshi commented on test/test_cuda.py:
```diff
@@ -22,6 +22,9 @@
 import torch.cuda
 import torch.cuda.comm as comm
+from torch.cuda._memory_viz import profile_plot
+from torch.cuda._memory_viz import trace_plot
```
Could you please add some brief comments in the function stubs about what profile_plot and trace_plot are? Also, a brief description of what this PR is changing the PR body would be helpful too.

zdevito replied:
yep, added docstrings and PR comment.

aaronenyeshi reviewed (changes requested):
Overall looks good, would like a little more comments in the _memory_viz.py tool.

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"

Adds a category legend to memory trace plots that colors allocations by their role (activation, parameter, gradient, etc.) as captured by kineto.




[ghstack-poisoned]

aaronenyeshi reviewed (approved):
Very nice! Thanks for the PR!

zdevito commented:
@pytorchbot merge

zdevito committed:
Update on "memory viz: Add colors for categories and a legend"

Adds a category legend to memory trace plots that colors allocations by their role (activation, parameter, gradient, etc.) as captured by kineto.




[ghstack-poisoned]

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43757381).

zdevito commented:
@pytorchbot merge

seemethere commented:
@pytorchbot revert -c ghfirst -m "PR is linked to the incorrect diff causing diff train import issues, see https://fburl.com/sandcastle/07oygj36"

seemethere commented:
I have no idea how this happened but D43757381 is somehow linked to this PR when it should actually be linked to https://github.com/pytorch/pytorch/pull/95541

seemethere commented:
Just reverted this manually with https://github.com/pytorch/pytorch/commit/b38b39c441f12be90fd5d7eafe74246d050665c8
